core/dom/range.js

```diff
@@ -2843,50 +2843,125 @@ CKEDITOR.dom.range = function( root ) {
-						rectArray;
+						rectList;
 
 					range.setStart( this.startContainer.$, this.startOffset );
 					range.setEnd( this.endContainer.$, this.endOffset );
+					var first,
+						textNode,
+						itemToInsertAfter;
 
-					rectArray = CKEDITOR.tools.array.map( range.getClientRects(), function( item ) {
-						return convertRect( item );
-					} );
+					rectList = range.getClientRects();
 
-					// In some cases ( eg. ranges contain only image ) IE will return empty rectList.
-					if ( !range.collapsed && !rectArray.length ) {
-						rectArray = [ getRect( this.createBookmark() ) ];
+					if ( !rectList.length ) {
+						// If native rect list is empty, use workarounds to get rects.
+
+						if ( !range.collapsed ) {
+							// In some cases ( eg. ranges contain only image ) IE will return empty rectList.
+
+							rectList = [ getRect( this.createBookmark() ) ];
+						} else if ( this.startContainer instanceof CKEDITOR.dom.element ) {
+							// If collapsed ranges are in element add textNode and return its rects.
+
+							first = this.checkStartOfBlock();
+							textNode = new CKEDITOR.dom.text( '\u200b' );
+
+							if ( first ) {
+								this.startContainer.append( textNode, true );
+							} else {
+								itemToInsertAfter = this.startContainer.getChildren().getItem( this.startOffset - 1 );
+								textNode.insertAfter( itemToInsertAfter );
+							}
+
+							// Create native collapsed ranges inside just created textNode.
+							range.setStart( textNode.$, 0 );
+							range.setEnd( textNode.$, 0 );
+
+							rectList = range.getClientRects();
+							textNode.remove();
+						} else if ( this.startContainer instanceof CKEDITOR.dom.text ) {
+							if ( this.startContainer.getText() === '' ) {
+								// In case of empty text fill it with zero width space.
+								this.startContainer.setText( '\u200b' );
+								rectList = range.getClientRects();
+
+								this.startContainer.setText( '' );
+
+							} else {
+								// If there is text node which isn't empty, but still no rects are returned use IE8 polyfill.
+								// This happens with selection at the end of line in IE.
+								rectList =  [ convertRect( getRect( this.createBookmark() ), isAbsolute, this ) ];
+							}
+						}
 					}
 
 					range.detach();
 
-					return rectArray;
+					return CKEDITOR.tools.array.map( rectList, function( item ) {
+						return convertRect( item, isAbsolute, this );
+					}, this );
 				};
 			} else {
-				return function() {
-					return [ getRect( this.createBookmark() ) ];
+				return function( isAbsolute ) {
+					return [ convertRect( getRect( this.createBookmark() ), isAbsolute, this ) ];
 				};
 			}
 
 			// Extending empty object with rect, to prevent inheriting from DOMRect, same approach as in CKEDITOR.dom.element.getClientRect().
-			function convertRect( rect ) {
+			function convertRect( rect, isAbsolute, context ) {
 				var newRect = CKEDITOR.tools.extend( {}, rect );
+
+				if ( isAbsolute ) {
```

Comment:
This is exactly the same logic as in `CKEDITOR.dom.element#getClientRects( true )`. I'm thinking about moving it into one place. Maybe to `CKEDITOR.dom.rect`? 🤔 @mlewand @engineering-this WDYT?

Reply:
Yup this looks like a copy'n'pasta so it should be extracted to a common place. I don't think that `CKEDITOR.dom.rect` is a good place though.

I'd start with a helper function for that, for instance in tools. As we'd get more members like that we can put them into a dedicated type, and deprecate old methods.